### PR TITLE
[JENKINS-59184] Add gitlab avatar cache

### PR DESCRIPTION
**Pull request by baymac**
Based on the issue reported by @ChristianCiach. See https://issues.jenkins-ci.org/browse/JENKINS-59184. I have added GitLab Avatar Cache (picked up from Gitea plugin). I am not sure if this fixes the issue he is currently facing as I do not have the resources to recreate the environment. I request you to test this and give your feedback.

### CODE REVIEW

**Comment by ChristianCiach** [comment]
Thanks. I will try to find the time to test it today. This depends on my workload though.

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/helpers/GitLabAvatar.java [hunk 1]
```diff
@@ -14,26 +13,31 @@ public GitLabAvatar(String avatar) {
         this.avatar = avatar;
     }
 
+//    @Override
+//    public String getAvatarImageOf(@NonNull String size) {
```
Comment (casz) [inline src/main/java/io/jenkins/plugins/gitlabbranchsource/helpers/GitLabAvatar.java]: Just remove it?

**Comment by casz** [comment]
@ChristianCiach you can download the HPI here: https://ci.jenkins.io/job/Plugins/job/gitlab-branch-source-plugin/job/feature%252Favatar-cache/1/artifact/io/jenkins/plugins/gitlab-branch-source/1.2.1-rc505.31270f7aa070/gitlab-branch-source-1.2.1-rc505.31270f7aa070.hpi

**Comment by baymac** [comment]
Yes, before merging I will remove that. Just keeping it there in case the
issue persists.

**Comment by ChristianCiach** [comment]
There is something funny happening here: Using the hpi file linked by @casz, the projects now have avatars in jenkins, but the avatars are not these from gitlab. Instead, they look a lot like autogenerated git**hub**-user avatars (like my avatar you are seeing to the left of this comment).

At first I thought these avatars are the git**lab**-user-avatars of the users owning the gitlab-projects, but this is not the case. The avatars seem to be generated by `GitLabAvatarCache.generateAvatar(String, int)`.

**Comment by ChristianCiach** [comment]
I think we should totally remove this code from the GitLabAvatarCache class and find out why it wasn't able to fetch the actual image. I will try to help you debug this, but it has been ages since I've last setup a jenkins debug environment. This could take a bit.

**Comment by ChristianCiach** [comment]
I don't know if this is the culprit, but we should probably add an `Authorization`-Header containing the API-Token to the HTTP call in `GitLabAvatarCache.FetchImage.call()`.

**Comment by baymac** [comment]
Can you wait for sometime until the image appears. Once the build is
complete, refresh the page you will see the avatar. At least this is how it
works on my machine.

**Comment by ChristianCiach** [comment]
Related gitlab issue: https://gitlab.com/gitlab-org/gitlab-ce/issues/54633

It looks like this issue is unsolvable at the moment. No matter how mighty your Access-Token is, you cannot access the image file using the "PRIVATE-TOKEN"-Header. Someone over there hinted about a possibility to access the raw image data using a new api endpoint that may be added sometime in the future. Until there is such a new api endpoint, I think this issue is unsolvable for now.

So, what should we do for now? I think we should come back to your earlier suggestion to provide a trait to disable the fetching of the project avatars. Fetching avatars works perfectly as-is (without this merge request) for public gitlab repos. But for people who are dealing with private repos, such a configurable trait could be a welcome feature.

**Comment by baymac** [comment]
So can we implement the trait to disable avatars for private repos only?

**Comment by ChristianCiach** [comment]
> So can we implement the trait to disable avatars for private repos only?

Sorry, this was badly worded on my part. "For private repos only" is not enough in our case. Our repos are all "public", but you still need to log into gitlab to see anything :) So, you shoudn't distinguish between public and private repos when implementing this trait.

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/helpers/GitLabAvatar.java (hunk 1 above)
Comment (markyjackson-taulia) [reply src/main/java/io/jenkins/plugins/gitlabbranchsource/helpers/GitLabAvatar.java]: I say yes

**Comment by casz** [comment]
This PR does not really solve the problem. The problem lies with GitLab 😢

**Comment by baymac** [comment]
@casz I am thinking to merge this with master as well. Reasons:

1. When offline avatar is still displayed
2. Self generates a dummy avatar when avatar cannot be accessed

**Comment by casz** [comment]
Not sure this avatar caching is doing a good job:

imageimage

**Pull request merged**